src/tools/non-promotion-reasons.md

```diff
@@ -111,16 +163,546 @@ one or more of the following:
 {{site.alert.end}}
 
 
+### Can't promote `this` {#this}
+
+**The cause:**
+You're trying to promote `this`,
+but type promotion for `this` is not yet supported.
+
+One common `this` promotion scenario is when writing [extension methods][].
+If the [`on` type][] of the extension method is a nullable type,
+you'd want to do a null check to see whether `this` is null:
+
+**Example:**
+
+{:.bad}
+```dart
+extension E on int? {
+  int? get i {
+    return this != null;
```

Comment:
This example doesn't actually trigger a type promotion failure. It's just a straight up ordinary compilation error because the result of a `!=` check is a `bool`, not an `int?`.

I would recommend a slightly different example, perhaps something like:
```dart
extension E on int? {
  int get valueOrZero {
    return this == null ? 0 : this;
  }
}
```

The corresponding "good" code would be:
```dart
extension E on int? {
  int get valueOrZero {
    final self = this;
    return self == null ? 0 : self;
  }
}
```

Reply:
Thank you! I tried simplifying the last example you gave me but messed it up in the process